### ISSUE TITLE
Code distorting in text (re: #218)

### ISSUE THREAD

**Issue by brideylea** [CoEDL/nyingarn-workspace#228]
Hi Conal, 

Submitting this bug on behalf of Gari. This is in reference to a previous Github issue (#218). When working on a transcription, errors are occurring in the code when highlighting text. In some instances clicking 'save' corrects the code, but other times the code is actually changing (not just superficially) and resulting in errors e.g. line 7 in the first image and line 8 in the second image. 

![Screenshot 2024-02-26 at 6 25 37 pm](https://github.com/CoEDL/nyingarn-workspace/assets/152234516/4bfa6116-6dd6-44c1-aaa4-7c86b8c9af1a)
![Screenshot 2024-02-25 at 9 36 36 pm](https://github.com/CoEDL/nyingarn-workspace/assets/152234516/acc8825c-17ed-4fd6-bab1-fde384b2587a)

See also images 3 and 4 for further instances. 

![Screenshot 2024-05-01 at 3 28 13 PM](https://github.com/CoEDL/nyingarn-workspace/assets/152234516/f0c8f05a-052d-45ed-ae66-4a6a31e3b5f5)
![Image 1-5-2024 at 4 58 PM](https://github.com/CoEDL/nyingarn-workspace/assets/152234516/08ce8a81-817b-4e4e-884f-dab2b2547148)

**Comment by gurridhala** [comment]
Is anyone working on this issue? It is still happening and I first posted the issue over 14 months ago.

**Comment by r-tae** [comment]
Thanks Bridey, I'll look into this; definitely my wheelhouse

This looks like a nasty bug, but the video in #218 looks helpful for diagnosing it

**Comment by r-tae** [comment]
@brideylea @gurridhala I've tried to reproduce this (I'm aslo MacOS), I've only been able to do so in Firefox; I can only make it happen maybe once every 10-15 tries and I can only do so by:

1. making a selection
2. dragging the selection (this is only possible in firefox) to move it somewhere else in the document
3. pressing cmd-z to undo and immediately selecting somewhere else in the document


Do you have another way that reproduces it more reliably? This is a bug in codemirror (or, at least, out use of codemirror) and that's a particularly complex library and at this rate I think it would take a long time for me to even find a cause, let alone fix it

**Comment by r-tae** [comment]
For my own future reference, I can reproduce it in a document with no soft line wraps.

![Image](https://github.com/user-attachments/assets/e774e3e6-b1e7-45a9-a58d-847cb4b975d3)

**Comment by r-tae** [comment]
I *can* reproduce it if I replace the entire document with a different one (valid xml) before testing; so its not related some weird bleedthrough of the initial state into the current working state.

If I put a line break in the middle of every `<line></line>` then it's reproducible, but is restricted to the end of every second line; meaning it follows the closing `</line>` tags.

It is reproducible if I rename every `<line>` tag to `<lime>` or `<a>`; *not* if I remove all `<line>`s; but if I then add a `<a />` to the end of every line its reproducible again

Very bizarre but at least I have some leads

**Comment by brideylea** [comment]
Very bizarre!! Thanks very much for investigating @r-tae. Could a potential short-term solution be for @gurridhala to try using a different browser if working in Firefox?

**Comment by r-tae** [comment]
@brideylea Yep, I think not using Firefox for now would be the move. Nick suggested we put a banner in the interface suggesting people don't use Firefox, so that may be forthcoming.

Soon I will also try bumping the versions of all the relevant libraries we use, that might get rid of the issue completely, who knows

**Comment by gurridhala** [comment]
@r-tae @brideylea Hi both, I've never used Firefox with Nyingarn - this is happening in Google Chrome. I'm going to drop two lots of screenshots with the error happening again, each lot is on one page in Nyingarn.
Also if you're trying to replicate errors, could you please do it in manuscripts that haven't been worked on (0 pages complete). If there are any transcription errors introduced into completed manuscripts it is likely to go unnoticed since we have exported them already, so the transcription error will end up in the repository.

**Comment by gurridhala** [comment]
<img width="1279" alt="Image" src="https://github.com/user-attachments/assets/a3f1130d-c4f4-46cf-a0f4-9b6386093695" />
<img width="1253" alt="Image" src="https://github.com/user-attachments/assets/cfe0b0f2-298c-4968-89a3-1aa1450bad74" />

Notice the code for /line goes white before it changes to the /b.

<img width="1271" alt="Image" src="https://github.com/user-attachments/assets/cfd06d2f-7c7c-43b6-8473-ffb26844f273" />
<img width="825" alt="Image" src="https://github.com/user-attachments/assets/da0f42a6-d8c7-4ca1-8b54-09997c630f3c" />
<img width="820" alt="Image" src="https://github.com/user-attachments/assets/565157f1-87fc-4661-a141-c157619d9821" />
<img width="829" alt="Image" src="https://github.com/user-attachments/assets/edd65003-92be-4c17-b609-3757c2d58804" />
<img width="826" alt="Image" src="https://github.com/user-attachments/assets/fce48c8f-8725-41ff-9779-29cdbb3d524d" />

**Comment by r-tae** [comment]
@gurridhala thanks for those extra details; unfortunately this makes the bug a dead end for me, I can't reproduce it at all on Chrome. The best I can do for now is update the version of the editing component we use, in the hope that it'll magically fix it.

The only other possibility is, if @brideylea and @nthieberger are amenable to it, to remove the library that I believe to be responsible; the impact of this is that the text would no longer be syntax highlighted (so the text would all be plain black-on-white) and I think the format button would have to be removed. This would make TEI syntax errors harder to spot because the colours help you notice when you're missing a `</line>` or something, but I think this bug is serious enough to warrant it

I have only tested on documents I've imported into staging; I try to limit my interactions with production in case I do mess up any transcriptions that are in progress/finished.

**Comment by gurridhala** [comment]
@r-tae thanks River. I know Tom was keen to change the interface to be more user friendly, so users aren't dealing with code. I think this is currently still a barrier to usage that community members are put off by that. If this could be implemented, then the syntax not being highlighted would be a non-issue, and I assume it'd be unlikely that you'd end up with missing characters or commands that would break the code ?

**Comment by r-tae** [comment]
A WYSIWYG interface for TEI is quite a large work package, I don't see that being done soon but I will discuss with Nick.
